app/utils/risk_index.py

```diff
@@ -0,0 +1,81 @@
+from typing import List, Optional
+
+from sqlalchemy.orm import Session
+import pandas as pd
+
+import crud
+import utils
+import uuid
+
+from models import PestModel
+
+
+def calculate_risk_index_probability(db: Session, weather_dataset_id: int, pest_models: List[PestModel], parameter: Optional[str] = None):
+    # SQL query for the data
+    data_db = crud.data.get_data_interval_query_by_dataset_id(db=db, dataset_id=weather_dataset_id)
+
+    df = pd.read_sql(sql=data_db.statement, con=db.bind, parse_dates={"date": "%Y-%m-%d"})
+
+    # Calculate the risks associated with each pest_model
+    for pm in pest_models:
+        risks_for_current_pm = ["Low"] * df.shape[0]
+
+        for rule in pm.rules:
+            conds = []
+            for cond in rule.conditions:
+                conds.append("(x['{}'] {} {})".format(cond.unit.name, cond.operator.symbol, float(cond.value)))
+            pre_final_str = ""
+            for cond_strs in conds[:len(conds) - 1]:
```

Comment:
@vlf-stefan-drobic If you want to iterate over list except last element _**for cond_strs in conds[:-1]**_ would do the job I think, no need for len operation.
Also this can be written with 2 loop not three  (just do the oposite logic, use first elem):
for rule in pm.rules:
    pre_final_str = rule.conditions[0] 
    for i, cond in rule.conditions:
          cond_str = "(x['{}'] {} {})".format(cond.unit.name, cond.operator.symbol, float(cond.value))
          pre_final_str = pre_final_str+ " & "+ cond_str

Reply:
Yeah, len operation is unnecessary as well as the temp list, I can just append them in the loop, updating with refactored code